DoSomething/gambit-conversations
# Conditionally include why_participated when creating photo posts

#### What's this PR do?

Modifies the under-construction `topics/posts/photo` middleware to only ask user for their signup` why_participated` if it currently doesn't exist in Rogue, completing feature parity for creating photo posts via the [Gambit Campaigns `/POST campaignActivity` endpoint](https://github.com/DoSomething/gambit-campaigns/tree/master/lib/middleware/campaignActivity/photo)

* Adds request helpers for getting and saving draftSubmission values

* Modifies user signup helpers to pass specific `campaign`, `signupSource`, and `signupSourceDetails` arguments instead of a single `args` argument

* Modifies user post helpers to pass specific `campaign`, `postSource`, and post value params arguments instead of a single `args` argument

* Adds middleware comments to detail multiple question flow per https://github.com/DoSomething/gambit-conversations/pull/445/files#r235509064

* Validates input for caption and whyParticipated with `helpers.util.isValidTextFieldValue` (renaming `helpers.util.isValidTextPost`)


#### How should this be reviewed?

* Verify signups are created via keyword (delete signup if exists already to confirm)
* Verify signup is fetched via keyword if signup exists already
* Verify that you're asked for why participated when creating your 1st photo post for the signup
* Verify that you are NOT asked for why participated when creating your 2nd photo post for the signup
* Verify caption and why participated are only accepted when message text has letters and is more than 2 characters (e.g. messages with only emoji, or just a photo should return an invalid response reply)

#### Any background context you want to provide?

This should be tested on staging, as we're looking to set the `DS_GAMBIT_CONVERSATIONS_DRAFT_SUBMISSIONS_ENABLED` config to `true` to go live.

Will need to investigate how many open draft submissions there are currently before this goes live to assess whether this can go live as is (or without an import to backfill existing draft submissions into our `draftSubmissions` collection.

#### Relevant tickets

https://www.pivotaltracker.com/story/show/161494958

#### Checklist
